### DISCUSSION TITLE
Dynamic Route Challenges: Preventing Unnecessary Bundle Loading

### DISCUSSION THREAD

**Discussion by haapti** [vercel/next.js#69563] [App Router]
### Summary
Im using dynamic routes in order to do internationalization. Some regions has region specific RSC components on a home page. Code splitting seems to work fine according to bundle analyzer results and all components goes into their own bundles. The issue is that all of these bundles gets loaded into client regardless of whether the components are in use on specific region.

### App dir structure

All routes are using prerendered as static HTML (SSG). Im not using dynamic rendering or streaming.

- layout.tsx
- [locale]
  - layout.tsx
  - page.tsx

### Page.tsx logic

```
if (region === "en") return <ComponentA />
if (region === "fi") return <ComponentB />
```
### Issue
`<ComponentA />` and `<ComponentB />` bundles are loaded into client on both `/en` and `/fi` dynamic routes.

### I have tried so far
- Dynamic importing https://nextjs.org/docs/app/building-your-application/optimizing/lazy-loading#importing-server-components
- Defining locales in generateStaticParams https://nextjs.org/docs/app/api-reference/functions/generate-static-params
- Suspense Boundaries

Not sure if it is possible to have Next.js to load only certain bundles into client depending on what is in dynamic route, even if you have pre-defined dynamic route segments beforehand.

**Comment by Zelenovskiy94** [comment]
I've faced the same issue 
 Any ideas to resolve it?

**Comment by rohitpotato** [comment]
@leerob Is this even possible, my team has been struggling with this for a quite a while now. We have a ecommerce site, which receives a lot of traffic daily (surpassing 5 million hits daily), and we want really want to leverage this feature. Is there a way, that the server itself can send the required bundle, because dynamic imports would mean another roundtrip from the client to the server.

Can you please take a look at this? I am not sure if there's already an open issue being tracked somewhere. We are on version `14.2.5`.

**Comment by oleksiimazurenko** [comment]
This is reachability, not chunking — and it's why the bundle analyzer misleads you here. Your `page.tsx` resolves `ComponentA`/`ComponentB` at runtime, so *statically* both are reachable from every `[locale]` route. The bundler can't prove `/en` never renders `ComponentB`, so it must ship both. `next/dynamic` doesn't help because its laziness is client-side — the server has already decided what to render before the client runs a byte; whatever *might* render must be reachable.

The fix that worked for us (SSG, ~700 pages, cut first-load JS 53%): move the "which components does this route use" knowledge from runtime to build time. A codegen step stamps one physical route file per page whose import map contains only that page's components, then `rewrites()` in `beforeFiles` maps the real URL to it. Since you already know your locales at build time via `generateStaticParams`, you can stamp one route per locale. Since it's build-time codegen, keep a fallback: if the manifest is missing, every URL just falls through to the original universal route — degraded, never broken.

Full writeup with the five dead ends (direct `import()`, file splitting, Turbopack config, webpack `splitChunks`) and the codegen + rewrites setup: https://oleksiimazurenko.dev/en/blog/nextjs-dynamic-imports-load-everything

**Comment by sireskandari** [comment]
The important part here is that conditional rendering and code splitting are two different things.

This:

```tsx
import ComponentA from './ComponentA'
import ComponentB from './ComponentB'

if (region === 'en') return <ComponentA />
if (region === 'fi') return <ComponentB />
```

prevents both components from being rendered, but both modules are still statically reachable from the route's module graph.

If these are Client Components and you genuinely want the browser to fetch only the region-specific implementation, I would move the selection behind a client-side dynamic import:

```tsx
'use client'

import dynamic from 'next/dynamic'

const ComponentA = dynamic(() => import('./ComponentA'))
const ComponentB = dynamic(() => import('./ComponentB'))

export function RegionComponent({
  region,
}: {
  region: 'en' | 'fi'
}) {
  if (region === 'en') {
    return <ComponentA />
  }

  if (region === 'fi') {
    return <ComponentB />
  }

  return null
}
```

`next/dynamic` creates separate client chunks and the component is loaded when that branch is actually rendered.

One subtlety with the App Router is important here: dynamically importing a Client Component from a **Server Component** currently does not give you the same automatic client code-splitting behavior. So if your current `page.tsx` is a Server Component, simply changing:

```tsx
import ComponentA from './ComponentA'
```

to:

```tsx
const ComponentA = dynamic(() => import('./ComponentA'))
```

at that level may not solve what you are measuring.

I would keep the locale/region determination on the server, pass the small `region` value to a Client Component, and let that Client Component own the dynamic imports.

Something like:

```tsx
// page.tsx - Server Component

import { RegionComponent } from './RegionComponent'

export default function Page({
  params,
}: {
  params: { locale: string }
}) {
  const region = params.locale === 'fi' ? 'fi' : 'en'

  return <RegionComponent region={region} />
}
```

Then `RegionComponent.tsx` contains the `next/dynamic` imports.

Also make sure the extra requests you're seeing aren't actually caused by route prefetching. Next.js automatically prefetches routes linked with `<Link>` in many cases, so DevTools can show chunks arriving even though the corresponding component wasn't rendered.

I'd test with:

1. Disable the browser cache.
2. Hard-refresh directly onto one locale.
3. Filter Network by JS.
4. Temporarily remove or set `prefetch={false}` on links to the other locale.
5. Confirm whether the other region's chunk is still fetched.

If it disappears after moving the dynamic imports into a Client Component, then the issue was the module boundary rather than the conditional itself.
